File: python/src/sde/provisioning.py
"""Prepare physical schema using client-owned provisioning connections before opening runtime."""

from __future__ import annotations

from collections.abc import Mapping
from typing import TYPE_CHECKING, cast

from .errors import MigrationRefused
from .generation import Fencable, check_map_project
from .groups import colocation_groups
from .model import LogicalModel
from .placement import PlacementMap

if TYPE_CHECKING:
    from .session import Engine


def prepare_schema(
    model: LogicalModel,
    placement: PlacementMap,
    engines: Mapping[str, Engine],
    *,
    project_id: str | None = None,
) -> None:
    """Apply the described schema and initial generation; never advance an existing generation.

    Credentials remain in these local adapters. The runtime is opened afterwards on separately
    configured connections, so it need not receive the provisioning role's ALTER capability.
    """
    if model.version != placement.model_version:
        raise MigrationRefused("schema preparation needs the model named by the placement map")
    local_project = check_map_project(placement, project_id)
    needed = sorted(
        {material.engine for spot in placement.groups.values() for material in spot.all()}
    )
    if any(name not in engines for name in needed):
        raise MigrationRefused("schema preparation is missing an engine named by the map")
    if placement.contract >= 4 and any(
        not callable(getattr(engines[name], "write_fence", None)) for name in needed
    ):
        raise MigrationRefused("schema preparation needs native write generations on every engine")
    for group in colocation_groups(model):
        spot = placement.placement_of(group.name)
        keys = {name: model.entity(name).key for name in group.members}
        for material in spot.all():
            engine = engines[material.engine]
            engine.ensure_schema(material.layout, keys=keys)
            if spot.write_epoch is not None:
                for table in sorted(material.layout.tables.values()):
                    assert local_project is not None
                    cast(Fencable, engine).write_fence(table, project_id=local_project).prepare(
                        spot.write_epoch
                    )

File: typescript/src/provisioning.ts
/** Prepare client-owned physical schema before opening runtime connections. */
import { compareCodePoints } from './canonical.js'
import { EngineError, MigrationRefused } from './errors.js'
import { checkMapProject, type Fencable } from './generation.js'
import { colocationGroups } from './groups.js'
import type { LogicalModel } from './model.js'
import { refuseFindings } from './physical.js'
import type { PlacementMap } from './placement.js'
import type { Engine } from './session.js'
import type { WatermarkStore } from './watermark.js'

export async function prepareSchema(model: LogicalModel, placement: PlacementMap,
  engines: Readonly<Record<string, Engine>>, options: { projectId?: string } = {}): Promise<void> {
  if (model.version !== placement.modelVersion) {
    throw new MigrationRefused('schema preparation needs the model named by the placement map')
  }
  const project = checkMapProject(placement, options.projectId)
  const needed = [...new Set(Object.values(placement.groups).flatMap((spot) =>
    [spot.source, ...spot.derived].map((material) => material.engine)))].sort()
  if (needed.some((name) => engines[name] === undefined)) {
    throw new MigrationRefused('schema preparation is missing an engine named by the map')
  }
  if (placement.contract >= 4 && needed.some((name) =>
    typeof (engines[name] as Engine & Partial<Fencable>).writeFence !== 'function')) {
    throw new MigrationRefused('schema preparation needs native write generations on every engine')
  }
  for (const group of colocationGroups(model)) {
    const spot = placement.groups[group.name]!
    const keys: Record<string, readonly string[]> = {}
    for (const name of group.members) keys[name] = model.entities.find((entity) => entity.name === name)!.key
    for (const material of [spot.source, ...spot.derived]) {
      const engine = engines[material.engine] as Engine & Fencable
      // Provisioning is where a person can act on a physical difference, so here it refuses.
      refuseFindings((await engine.ensureSchema(material.layout, { keys })) ?? [], EngineError)
      if (spot.writeEpoch !== undefined) {
        for (const table of Object.values(material.layout.tables).sort()) {
          await engine.writeFence(table, { projectId: project as string }).prepare(spot.writeEpoch)
        }
      }
    }
  }
  if (placement.signed) {
    // Prepare storage without adopting this map's version before runtime activation.
    for (const name of Object.keys(engines).sort(compareCodePoints)) {
      const store = engines[name] as Engine & Partial<WatermarkStore>
      if (typeof store.mapWatermark === 'function' && typeof store.recordMapVersion === 'function') {
        await store.mapWatermark()
      }
    }
  }
}
